alf/environments/metadrive/extra_rewards.py

```diff
@@ -0,0 +1,348 @@
+    """Produce a cost based on the (harsh) brake.
+
+    The function is a squared cost of the violation that only activate when the
+    lon acceleration is beyond the harsh brake limit and the speed is also above
+    its deadband.
+
+    Also note that the cost is always POSITIVE or zero.
+
+    Args:
+
+        lon_acc: the longitudinal acceleration
+        speed: the value of the speed
+        harsh_brake_limit: the threshold used to determine whether the lon_acc
+            is considered a harsh brake to be penalized.
+        speed_deadband: cost is 0.0 if speed is below this threshold
+        scale: the scale of the squared violation
+        cap: if the scaled squared violation will be capped by this value
+
+    """
+    if lon_acc > harsh_brake_limit or speed < speed_deadband:
+        return 0.0
+
+    diff = harsh_brake_limit - lon_acc
+    return min(diff * diff * scale, cap)
+
+
+@alf.configurable
+class EgoKinematicReward(ExtraReward):
+    """The comfort rewards that are based on the kinematics of the ego vehicle.
+
+    Args:
+
+        harsh_brake_cost_func: the function that converts lon acc to a reward
+        lon_jerk_cost_func: the function that converts lon jerk to a reward
+        lat_jerk_cost_func: the function that converts lat jerk to a reward
+
+    """
+
+    def __init__(self,
+                 harsh_brake_cost_func=squared_brake_cost,
+                 lon_jerk_cost_func=squared_jerk_cost,
+                 lat_jerk_cost_func=squared_jerk_cost):
+        super().__init__()
+        self._harsh_brake_cost_func = harsh_brake_cost_func
+        self._lon_jerk_cost_func = lon_jerk_cost_func
+        self._lat_jerk_cost_func = lat_jerk_cost_func
+
+        # Stores the history of the linear speed and heading for computing the
+        # derivatives.
+        self._history_speed = np.zeros(3)
+        self._history_heading = np.zeros(5)
+
+    def evaluate(self, engine: BaseEngine):
+        ego: BaseVehicle = engine.managers["agent_manager"].active_agents[
+            "default_agent"]
+
+        self._history_heading = np.roll(self._history_heading, -1)
+        self._history_heading[-1] = ego.heading_theta
+
+        self._history_speed = np.roll(self._history_speed, -1)
+        self._history_speed[-1] = ego.speed / 3.6  # km/h -> m/s
+
+        lon_acc = _estimate_derivative_1(self._history_speed)
+        lon_jerk = _estimate_derivative_2(self._history_speed)
+        lat_acc = _estimate_derivative_2(self._history_heading, is_angle=True)
+        lat_jerk = _estimate_derivative_3(self._history_heading, is_angle=True)
+
+        speed = self._history_speed[-1]
+        lon_acc_cost = self._harsh_brake_cost_func(
+            lon_acc=lon_acc, speed=speed)
+        lon_jerk_cost = self._lon_jerk_cost_func(jerk=lon_jerk, speed=speed)
+        lat_jerk_cost = self._lat_jerk_cost_func(jerk=lat_jerk, speed=speed)
+
+        harsh_brake = 1.0 if speed > 2.0 and lon_acc < -1.2 else 0.0
```

Comment:
the const value 2.0 and -1.2 is also used as the default argument to harsh_break_cost_func. This may lead to some inconsistency.

Reply:
Right this is actually a bit awkward. You are right about different intentions here

1. Harsh brake cost function is the means to encourage mild braking instead of harsh braking
2. Harsh brake boolean is to judge whether a harsh brake happens or not.

I have updated the code logic here to make it more clear.